barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy.test.cpp

```diff
@@ -332,6 +332,73 @@ template <typename Flavor> class ProtoGalaxyTests : public testing::Test {
         decide_and_verify(prover_accumulator_2, verifier_accumulator_2, true);
     }
 
+    /**
+     * @brief Testing two valid rounds of folding followed by the decider for a structured trace.
+     *
+     */
+    static void test_full_protogalaxy_structured_trace()
+    {
+        bool structured = true;
+        TupleOfInstances instances = construct_instances(2, structured);
+
+        auto [prover_accumulator, verifier_accumulator] = fold_and_verify(get<0>(instances), get<1>(instances));
+        check_accumulator_target_sum_manual(prover_accumulator, true);
+
+        TupleOfInstances instances_2 = construct_instances(1, structured); // just one set of prover/verifier instances
+
+        auto [prover_accumulator_2, verifier_accumulator_2] = fold_and_verify(
+            { prover_accumulator, get<0>(instances_2)[0] }, { verifier_accumulator, get<1>(instances_2)[0] });
+        check_accumulator_target_sum_manual(prover_accumulator_2, true);
+        info(prover_accumulator_2->proving_key.circuit_size);
+        decide_and_verify(prover_accumulator_2, verifier_accumulator_2, true);
+    }
+
+    /**
+     * @brief Testing two valid rounds of folding followed by the decider for a structured trace.
+     * @details Here we're interested in folding inhomogeneous circuits, i.e. circuits with different numbers of
+     * constraints, which should be automatically handled by the structured trace
+     *
+     */
+    static void test_full_protogalaxy_structured_trace_inhomogeneous_circuits()
+    {
+        bool structured = true;
+
+        // Construct three circuits to be folded, each with a different number of constraints
+        Builder builder1;
+        Builder builder2;
+        Builder builder3;
+        // Create inhomogenous circuits by calling the circuit gen function multiple times
+        construct_circuit(builder1);
```

Comment:
It would probably be more readable if we just add the number of gates  as a parameter to construct_circuit, these for loops are a bit unreadable

Reply:
fait enough. I'll just rework it so it's only adding arithmetic gates which still gets the point across and is a bit more explicit

Reply:
all of the mock circuit stuff could use an overhaul